util/secscan/blob.py

```diff
@@ -0,0 +1,70 @@
+from urlparse import urljoin
+
+from flask import url_for
+
+from util.security.registry_jwt import generate_bearer_token, build_context_and_subject
+
+
+def _repository_and_namespace(repository_ref):
+    repo_name = repository_ref.name
+    namespace_name = repository_ref.namespace_name
+    return "/".join([namespace_name, repo_name])
+
+
+class BlobURLRetriever(object):
+    """
+    Helper class which encodes the logic for retrieving blobs for security indexing.
+    """
+
+    def __init__(self, storage, instance_keys, app):
+        assert storage is not None
+        assert instance_keys is not None
+        assert app is not None
+
+        self._storage = storage
+        self._instance_keys = instance_keys
+        self._app = app
+
+    def url_for_download(self, repository_ref, blob):
+        """
+        Returns the URL for downloading the given blob under the given repository.
+        """
+        # Try via direct download from storage.
+        uri = self._storage.get_direct_download_url(self._storage.locations, blob.storage_path)
+        if uri is not None:
+            return uri
+
+        # Otherwise, use the registry blob download endpoint.
+        with self._app.test_request_context("/"):
+            return url_for(
```

Comment:
Clair returns this error:
```
"failed to fetch layers: encountered error while fetching a layer: fetcher: request failed: Get /v2/alecmerdler/bad-image/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4: unsupported protocol scheme \\"\\""
```

The protocol and hostname are missing, only leaving the `/v2/...` path. Also, I'm not sure why you return here but then have another return after with `urljoin()`, which is using undefined values.